Add bounds arg to {}.scope

prko opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation
This PR is a part of #6230

<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->

## Types of changes

<!-- Delete lines that don't apply -->

- Documentation
- New feature

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

prko commented:
@mtmccrea 

Hi, I think I split #6230 as you advised.
However, `{}.scope(bounds: ... )` in this PR does not work because of the following line:

https://github.com/supercollider/supercollider/blob/96eafe34c66fc76e94e3a9016f77a3f2cf6b341e/SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc#L48

This line should be changed as follows:
```
server.scope(numChannels, outbus, bufsize, zoom, outUGen.rate). bounds_(bounds); 
```
However, #6181 should be approved and merged first.
Um... 
Should I wait for #6281 to be merged, or should I change it now?

mtmccrea commented:
> However, https://github.com/supercollider/supercollider/issues/6181 should be approved and merged first.

That's fine, you can just make a note of that in the PR description above, and we can put the `has merge note` tag on this.

prko commented:
Thanks for `adding has merge note`.
>is Make s.plotTree a singleton window related to or required for setting its bounds? 

Making s.plotTree a singleton window is a separate issue from adding the bounds argument.

Currently SuperCollider makes s.plotTree every time it evaluates `s.plotTree`. I think this behaviour should be changed, and I changed it when I added the bounds argument. It is technically in a different category, so it should be written with a separate PR. However, I thought it was better to make one complex PR rather than many PRs with a tiny change. 

If you like it, I will split it. 

After splitting it again, the original PR is split into 5 PRs. 

(
Technically, this PR should be split into three: 
1. add bounds arg to {}.scope; 
3. add bounds arg to s.plotTree; 
4. 3. make s.plotTree a singleton window... 

Oh... the thing I thought was one is actually 6 things. Sorry again, I misunderstood submitting PR when I did it
)

mtmccrea commented:
>Making s.plotTree a singleton window is a separate issue from adding the bounds argument.

Then it should probably be a different PR

prko commented:
Thanks! I will do it!

prko commented:
@mtmccrea 
>Then it should probably be a different PR

I spilt only making `.plotTree` singleton Window from this PR.
It has now one more _Merge Note_:
When this PR is approved or merged, the following PR should be checked to see if each of them has already been approved or merged.
- #6294 
- #6281 (This should be approved and merged before this PR.)

prko commented:
I have removed the part related to `.plotTree' as this is covered in #6294.
I hope this removal will speed up the review of this PR.

telephon commented on HelpSource/Classes/Function.schelp:
```diff
@@ -510,7 +510,8 @@ argument::bufsize
 argument::zoom
 A zoom value for the scope's X axis. Larger values show more. The default is 1.
-
+argument:: bounds
```
maybe you don't need to add this long explanation, because the behaviour of the scope view may change. Maybe just refer to the documenattion of the scope view.

prko replied:
Thank you for that. 

However, `ScopeView` has no corresponding part in the documentation and `link::Classes/View#*new::` is linked from the inherited class methods.
Also, the `bounds' argument is related to the bounds method of the `Stethoscope'.

Which of the following should be placed in line 514?

1. `The bounds of the scope window. The default is nil and the scope window will be in the centre of the display. See link::Classes/Stethoscope#-bounds::`
2. `The bounds of the scope window. The default is nil and the scope window will be in the centre of the display. See link::Classes/View#*new::`

telephon replied:
It is the bounds of the stethoscope, right?

prko replied:
Yes. 

I originally proposed it as an argument of the Stethoscope class: 
    #6230 

Then @mtmccrea suggested it as a method: 
     https://github.com/supercollider/supercollider/pull/6230#issuecomment-2068195977

I agreed: 
    https://github.com/supercollider/supercollider/pull/6230#issuecomment-2068293011

mtmccrea replied:
>1. The bounds of the scope window. The default is nil and the scope window will be in the centre of the display. See link::Classes/Stethoscope#-bounds::

This is probably most appropriate because it calls `Stethoscope:-bounds`

It may clarify more to write:|
"The bounds of the scope code::Window::."

This would disambiguate it from the embedded ScopeView.

prko replied:
Thanks, done: https://github.com/supercollider/supercollider/pull/6282/commits/9c98ba7f047645fc92b4e9ecda8cdac777061448

telephon reviewed (changes requested):
good, just reduce documentation :)

prko commented on SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc:
```diff
@@ -45,7 +45,7 @@
 		synthMsg = synth.newMsg(server, [\i_out, outbus, \out, outbus], \addToHead);
 		server.sendMsg("/d_recv", bytes, synthMsg);
-		server.scope(numChannels, outbus, bufsize, zoom, outUGen.rate);
+		server.scope(numChannels, outbus, bufsize, zoom, outUGen.rate).bounds_(bounds); 
```
I have found a problem when using this PR's change: #6674
This issue occurs due to the following part:
```
.bounds_(bounds);
```
in line 48. I have no idea how to fix it.

prko commented:
I have found a problem using this PR. See the following comment on line 48 of the server-scope.sc file:
https://github.com/supercollider/supercollider/pull/6282/files#r2008764556

prko commented:
This PR is now independent of all other PRs and works fine.

telephon commented on SCClassLibrary/Common/GUI/tools/Stethoscope.sc:
```diff
@@ -437,8 +437,9 @@ Stethoscope {
-	
+
 	bounds_ { arg rect;
+		rect = rect ?? view.bounds;
```
should be 
`rect = rect ?? { view.bounds };`

prko replied:
Thanks! Done!
